core/lib/Cardano/Multisig.hs

```diff
@@ -0,0 +1,128 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TypeApplications #-}
+
+module Cardano.Multisig
+    (
+      MultisigScript (..)
+    , VerificationKeyHash (..)
+    , fromVerificationKey
+    , verificationKeyHashFromBytes
+    , toScriptHash
+
+    -- * Internal
+    , toCBOR
+    ) where
+
+import Prelude
+
+import Cardano.Address
+    ( invariantSize )
+import Cardano.Address.Derivation
+    ( Depth (..), XPub )
+import Cardano.Address.Style.Shelley
+    ( Shelley, blake2b224, pubkeyHashSize )
+import Control.DeepSeq
+    ( NFData )
+import Crypto.Hash
+    ( hash )
+import Crypto.Hash.Algorithms
+    ( Blake2b_224 )
+import Data.ByteString
+    ( ByteString )
+import Data.Foldable
+    ( foldl' )
+import Data.Word
+    ( Word8 )
+import GHC.Generics
+    ( Generic )
+
+import qualified Cardano.Codec.Cbor as CBOR
+import qualified Codec.CBOR.Encoding as CBOR
+import qualified Data.ByteArray as BA
+
+
+-- | A 'MultisigScript' type represents multi signature script. The script embodies conditions
+-- that need to be satisfied to make it valid.
+--
+-- @since 3.0.0
+data MultisigScript =
+      RequireSignatureOf VerificationKeyHash
+    | RequireAllOf [MultisigScript]
+    | RequireAnyOf [MultisigScript]
+    | RequireMOf Word8 [MultisigScript]
+    deriving stock (Generic, Show, Eq)
```

Comment:
Scripts are going to be a bit more general than multisig. Shall I suggest simply calling this `Script` ? Or `TransactionScript` ?

Reply:
agree, `Script` is better -> adopted in https://github.com/input-output-hk/cardano-addresses/pull/76/commits/84b8ea729080d3ac8e452c5304bbadd545f497af